docs/howto_contribute.md

```diff
@@ -0,0 +1,396 @@
+could look like this:
+
+```yaml
+---
+- name: Fail if required variables are not set
+  fail:
+    msg: "The required variables are not set."
+  when: image_name is not defined or
+        build_path is not defined or
+        dockerfile_name is not defined
+
+- name: Build Docker image
+  command: "docker build -t {{ image_name }} -f {{ build_path }}/{{ dockerfile_name }} {{ build_path }}"
+
+- name: Verify Docker image was built
+  command: docker images
+  register: docker_images
+  failed_when: "'{{ image_name }}' not in docker_images.stdout"
+```
+
+This completely simplifies the role to handle the building of a Docker image
+when given the required parameters.  In this simplified version, you would
+have to have separate roles/tasks that would handle the creation of any
+required directories and copying the Dockerfile to the host.
+
+### Writing Your First Test
+
+Now you have a new role created, let's put it to use in a new test playbook.
+
+#### Directory Setup
+
+Firstly, you will have to create a new sub-directory under `tests` and create
+symlinks to the top-level `callback_plugins` and `roles` directories.  This
+allows the test to reference the available roles and have its output nicely
+formatted by the callback plugin.  In this example, we are going to pretend we
+are creating a test playbook for the `docker build` command.
+
+```bash
+$ mkdir -p tests/docker-build/
+$ ln -s ../../callback_plugins/ tests/docker-build/callback_plugins
+$ ln -s ../../roles/ tests/docker-build/roles
+```
+
+Additional directories such as `files`, `templates` or `vars` may be required
+for your test.  Please reference the [Ansible documentation](http://docs.ansible.com/ansible/playbooks_best_practices.html#content-organization) about
+content organization about when and how to use these directories.
+
+#### Playbook Structure + System State
+
+Now we can start developing the actual playbook.  When running an Ansible
+playbook, the operations are mostly executed in serial, so there is a sense
+of state of the host under test.  This means you can structure your playbook
+to assume certain conditions of the host during the execution of the playbook.
+
+A simple playbook will have a list of tasks or roles defined in a single file
+and that is all that will be required.  Sometimes, it is desirable to create
+multiple playbooks in a single file for the purposes of providing delineation
+between certain features being tested.  This can be seen in [tests/admin-unlock](/tests/admin-unlock/main.yml)
+where each set of features is separated into their own playbooks.
+
+For our example, we will use the simple case of a single playbook in a single
+file.
+
+#### Playbook Sections
+
+All of the playbooks should start with a `name`, `hosts`, and `become`
+declaration.  The name is up to the test developer, but we almost always use
+`hosts: all` and `become: yes`.  The `hosts: all` allows the executor of the
```

Comment:
The `default()` filter is a good suggestion.  We've been able to get away with `hosts: all` thusfar because the focus has been single node tests.

I'll investigate updating our `hosts:` declaration across our tests and see how it works for us.